### PR TITLE
Issues 1966

### PR DESCRIPTION
**Pull request by raftaar1191**
## Description
PR for featured #1966 

## How Has This Been Tested?
Manually Tested

## Video Link (jpeg or gifs if applicable):
http://screencast-o-matic.com/watch/cbjFrmlM1X

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style.
- [ ] My code follows has proper inline documentation.

### CODE REVIEW

**Comment by raftaar1191** [comment]
@ravinderk  @DevinWalker 

Please review this PR and merge 

For more information on the PR please link on the below video link http://screencast-o-matic.com/watch/cbjFrmlM1X

**Comment by ravinderk** [comment]
@raftaar1191 please test your pr with lowest and highest supported version of PHP when you ask to review PR.
Getting this error `PHP Fatal error:  Can't use method return value in write context in /app/public/wp-content/plugins/Give/includes/misc-functions.php on line 1527` on PHP`5.4`

File: assets/js/admin/admin-scripts.js [hunk 1]
```diff
@@ -2190,3 +2196,72 @@ function get_url_params() {
...
+
+/**
+ * Upload CSV ajax
+ *
+ * @since 1.8.13
+ */
+function give_on_donation_import_ajax() {
+    var $form = jQuery('form.tools-setting-page-import');
+
+    var progress = jQuery( $form ).find( '.give-progress' );
```
Comment (ravinderk) [inline assets/js/admin/admin-scripts.js]: @raftaar1191 No need to wrapper `$form` again with`jQuery`. It is already a `jQuery` object

File: assets/js/admin/admin-scripts.js [hunk 2]
```diff
@@ -2190,3 +2196,72 @@ function get_url_params() {
...
+    var total_ajax = jQuery( progress ).data( 'total_ajax' );
```
Comment (ravinderk) [inline assets/js/admin/admin-scripts.js]: @raftaar1191 Do not need to add `var` for each variable, use javascript multi-line variable declaration structure

File: assets/js/admin/admin-scripts.js [hunk 3]
```diff
@@ -2190,3 +2196,72 @@ function get_url_params() {
...
+    var end = jQuery( progress ).data( 'end' );
+    var next = jQuery( progress ).data( 'next' );
+    var total = jQuery( progress ).data( 'total' );
+    var per_page = jQuery( progress ).data( 'per_page' );
+
+    jQuery.ajax({
+        type: 'POST',
+        url: ajaxurl,
+        data: {
+            action: give_vars.give_donation_import,
```
Comment (ravinderk) [inline assets/js/admin/admin-scripts.js]: @raftaar1191 Why we need dynamic action for this process? Can we hardcode this?

File: assets/js/admin/admin-scripts.js [hunk 4]
```diff
@@ -2190,3 +2196,72 @@ function get_url_params() {
...
+            start: start,
+            end: end,
+            next: next,
+            total: total,
+            per_page: per_page,
+            fields: $form.serialize()
+        },
+        dataType: 'json',
+        success: function ( response ) {
+            console.log( response );
```
Comment (ravinderk) [inline assets/js/admin/admin-scripts.js]: @raftaar1191 Please remove `console.log` after testing

File: includes/admin/admin-actions.php [hunk 5]
```diff
@@ -388,11 +388,120 @@ function _give_show_test_mode_notice_in_admin_bar( $wp_admin_bar ) {
 		'title'  => esc_html__( 'Give Test Mode Active', 'give' ),
 		'meta'   => array( 'class' => 'give-test-mode-active' ),
 	) );
-
 	return true;
 }
 add_action( 'admin_bar_menu', '_give_show_test_mode_notice_in_admin_bar', 1000, 1 );
 
+
+function give_import_page_link_callback() {
```
Comment (ravinderk) [inline includes/admin/admin-actions.php]: @raftaar1191 Add code doc block to this function

File: includes/admin/admin-actions.php [hunk 6]
```diff
@@ -388,11 +388,120 @@ function _give_show_test_mode_notice_in_admin_bar( $wp_admin_bar ) {
...
+	?>
+    <a href="<?php echo esc_url( give_import_page_url() ); ?>"
+       class="page-import-action page-title-action"><?php esc_html_e( 'Import Donations', 'give-manual-donations' ); ?></a>
```
Comment (ravinderk) [inline includes/admin/admin-actions.php]: @raftaar1191 Fix text domain

File: includes/admin/admin-actions.php [hunk 7]
```diff
@@ -388,11 +388,120 @@ function _give_show_test_mode_notice_in_admin_bar( $wp_admin_bar ) {
...
+	<?php
+
+	// Check if view donation single page only.
+	if ( ! empty( $_REQUEST['view'] ) && 'view-payment-details' === (string) sanitize_text_field( $_REQUEST['view'] ) && 'give-payment-history' === (string) sanitize_text_field( $_REQUEST['page'] ) ) {
```
Comment (ravinderk) [inline includes/admin/admin-actions.php]: @raftaar1191 Use `give_clean`

File: includes/admin/admin-actions.php [hunk 8]
```diff
@@ -388,11 +388,120 @@ function _give_show_test_mode_notice_in_admin_bar( $wp_admin_bar ) {
...
+
+
+function give_donation_import_callback() {
+	$fields = isset( $_POST['fields'] ) ? $_POST['fields'] : null;
+	parse_str( $fields );
+
+	// Parent key id.
+	$main_key = maybe_unserialize( $main_key );
+
+	$current    = (int) sanitize_text_field( $_REQUEST['current'] );
```
Comment (ravinderk) [inline includes/admin/admin-actions.php]: @raftaar1191 use `absint` instead

File: includes/admin/admin-actions.php [hunk 9]
```diff
@@ -388,11 +388,120 @@ function _give_show_test_mode_notice_in_admin_bar( $wp_admin_bar ) {
...
+	if ( function_exists( 'give_payment_save_page_data' ) ) {
+		add_action( 'give_insert_payment', 'give_payment_save_page_data' );
+	}
+	add_action( 'give_insert_user', 'give_new_user_notification', 10, 2 );
+	add_action( 'give_complete_donation', 'give_trigger_donation_receipt', 999 );
+
+	if ( $next == false ) {
+		$json_data = array(
+			'success' => true,
+			'message' => __( 'All donation uploaded successfully!', 'give-manual-donations' ),
```
Comment (ravinderk) [inline includes/admin/admin-actions.php]: @raftaar1191 Fix text domain

File: includes/admin/tools/views/html-admin-page-imports.php [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+if ( ! current_user_can( 'manage_give_settings' ) ) {
+	return;
+}
+
+/**
+ * Fires before displaying the import div tools.
+ *
+ * @since 1.5
```
Comment (ravinderk) [inline includes/admin/tools/views/html-admin-page-imports.php]: @raftaar1191 Why @since tag set to `1.5`

File: includes/process-donation.php [hunk 11]
```diff
@@ -895,8 +897,8 @@ function give_get_donation_form_user( $valid_data = array() ) {
 	$user['address']['line2']   = ! empty( $_POST['card_address_2'] ) ? sanitize_text_field( $_POST['card_address_2'] ) : false;
 	$user['address']['city']    = ! empty( $_POST['card_city'] ) ? sanitize_text_field( $_POST['card_city'] ) : false;
 	$user['address']['state']   = ! empty( $_POST['card_state'] ) ? sanitize_text_field( $_POST['card_state'] ) : false;
-	$user['address']['country'] = ! empty( $_POST['billing_country'] ) ? sanitize_text_field( $_POST['billing_country'] ) : false;
 	$user['address']['zip']     = ! empty( $_POST['card_zip'] ) ? sanitize_text_field( $_POST['card_zip'] ) : false;
+	$user['address']['country'] = ! empty( $_POST['billing_country'] ) ? ( ( $country_code = array_search( sanitize_text_field( $_POST['card_zip'] ), give_get_country_list() ) ) ? $country_code : sanitize_text_field( $_POST['card_zip'] ) ) : false;
```
Comment (ravinderk) [inline includes/process-donation.php]: @raftaar1191 use `give_clean`

File: assets/js/admin/admin-scripts.js [hunk 12]
```diff
@@ -2190,3 +2196,72 @@ function get_url_params() {
...
+            } else {
+                if( response.success == true ) {
+                    window.location = response.url;
+                } else {
+                    window.location = response.url;
+                }
+            }
+        },
+        error: function () {
+            alert( give_vars.error_message );
```
Comment (ravinderk) [inline assets/js/admin/admin-scripts.js]: @raftaar1191 are you sure this will show an error message?

File: assets/js/admin/admin-scripts.js (hunk 1 above)
Comment (raftaar1191) [inline assets/js/admin/admin-scripts.js]: Solved

File: assets/js/admin/admin-scripts.js (hunk 2 above)
Comment (raftaar1191) [inline assets/js/admin/admin-scripts.js]: Solved thanks

File: assets/js/admin/admin-scripts.js (hunk 4 above)
Comment (raftaar1191) [inline assets/js/admin/admin-scripts.js]: Remove thanks

File: assets/js/admin/admin-scripts.js (hunk 12 above)
Comment (raftaar1191) [inline assets/js/admin/admin-scripts.js]: yes it's working

File: includes/admin/admin-actions.php (hunk 5 above)
Comment (raftaar1191) [inline includes/admin/admin-actions.php]: Done thanks

File: includes/admin/admin-actions.php (hunk 6 above)
Comment (raftaar1191) [inline includes/admin/admin-actions.php]: Done thanks

File: includes/admin/admin-actions.php (hunk 7 above)
Comment (raftaar1191) [inline includes/admin/admin-actions.php]: Done thanks

File: includes/admin/admin-actions.php (hunk 8 above)
Comment (raftaar1191) [inline includes/admin/admin-actions.php]: Done thanks

File: includes/admin/admin-actions.php (hunk 9 above)
Comment (raftaar1191) [inline includes/admin/admin-actions.php]: Done thanks

File: includes/admin/tools/views/html-admin-page-imports.php (hunk 10 above)
Comment (raftaar1191) [inline includes/admin/tools/views/html-admin-page-imports.php]: Done thanks

File: includes/process-donation.php (hunk 11 above)
Comment (raftaar1191) [inline includes/process-donation.php]: Done thanks

File: assets/js/admin/admin-scripts.js (hunk 3 above)
Comment (raftaar1191) [inline assets/js/admin/admin-scripts.js]: Yes we can there is not differences

**Comment by DevinWalker** [comment]
@raftaar1191 change request:

**Comment by DevinWalker** [comment]
@raftaar1191 a few more items:

## 1. WP User Creation

Under what criteria is a WP user created? Some people may or may not want to create WP users for their donors. I'd like to discuss the best approach here for:

1. Creating _new_ WP users for new imported donors - I was thinking maybe adding an option during the import process like "Create WP users for new donors?"  
2.  Associating _existing_ WP users for new imported donors (possibly by providing an existing WP User ID via a column in the CSV).

## 2. Repeated Imports and Duplicates

I noticed that I am able to keep importing over and over again. It appears that the donors and forms are not duplicate, which is good. However, the _donations_ are created. Is this intentional? If an imported donation happened on the _same_ date and time it shouldn't be imported. That's a duplicate. 

## 3. Final Import "Report"

When you are done importing the breakdown isn't very descriptive. There is a lot more information that could be listed:

1. Number of duplicate donors, forms, wp users, and donations detected and skipped
2. The number of donors, forms, wp users, and donations created

Please update so that the final report is more descriptive so admins have a better idea what imported, what was skipped, and why.
 


## 4. Test vs Live Donations

Currently there is no way the I can see to specify whether an imported donation has been made in LIVE or TEST mode. I have added a column to the Google Doc spread sheet which includes a new "Test Mode" column that is boolean.  If this column does _not_ exist within the import the admin should still be able to specify whether the import is should be in test or live mode.



## 5. New Donors Donations Count

This appears to be an issue with breaking the cache or an improper import that needs fixing:



## 6. Remove the "Imported" badge and Add "Imported on ..." Note

Please remove this badge here:



And in its place add a payment note: "This donation was imported by admin@website.com" that should be the _first_ note for the payment:



Also, for any _donor_ that is imported there also need to be a note:



## 7. Empty Note Created on Import

I noticed that the imported is creating a note even if the note is blank:



## 8. CSS Issue with "Import Donations" button

Deactivate the "Manual Donations" add-on and then see the CSS issue here:



## 9. Delete Imported Data Tool Not Deleting All Data

I know our users are going to try this out and some will want to remove the imported data. The tool you've built seems to delete _some_ of the data but some still remains after running:



## 10. Typos to Fix







2017-08-22_14-46-43 copy



---

Please update the imported with the aforementioned issues. Ping me if you have any questions. Thanks!

**Comment by DevinWalker** [comment]
How's it going on this one @raftaar1191 can I get an update?

**Comment by raftaar1191** [comment]
Solved all the Changes request from @ravinderk 


Related to all your change request 
1. WP User Creation
In the CSV there is an option to enter Donor ID and Donor Email id both. If donor id is not empty then it will search for that ID in the user list. If found then all the details will be fetched from that user id or if that id does not exist then it will search for the email if that email is already registered then all the details from that email will be fetch and if that email is not being registered then it will register a new user with that email followed by first name and last name.


2. Repeated Imports and Duplicates
Solved, Now where a new donation is imported it will first check if this donation already exists or not. It will first check with the same Form id, price, author, and date as well. If all the match then it assumes to be the same donation and if not then 

3. Final Import "Report"
Working on it

4. Test vs Live Donations
Now there is a column in the CSV to mention that the donation is ( test or live ) donation. The value should in a boolean value. If in the column no found then there is an option on the first stepto choose the type of donation 

5. New Donors Donations Count
Working on it

6. Remove the "Imported" badge and Add "Imported on ..." Note
Done, Added in donation single page as well in donor notes section 

7. Empty Note Created on Import
Solved 

8. CSS Issue with "Import Donations" button
Solved

9. Delete Imported Data Tool Not Deleting All Data
Working on it

10. Typos to Fix
Fixed


11. Media Uploader Modal
Working on it

**Comment by raftaar1191** [comment]
@DevinWalker  
Related to you comment point number 
**11. Media Uploader Modal**
->It will be better to used ID in place of URL

File: assets/js/admin/admin-scripts.js [hunk 13]
```diff
@@ -1340,11 +1346,18 @@ jQuery.noConflict();
 						$media_modal_config = {
 							title   : give_vars.metabox_fields.file.button_title,
 							button  : {text: give_vars.metabox_fields.file.button_title},
-							multiple: false
+							multiple: false,
+                            contentUserSetting :false,
```
Comment (ravinderk) [inline assets/js/admin/admin-scripts.js]: @deepak can you explain this var?

File: includes/admin/admin-actions.php [hunk 14]
```diff
@@ -391,9 +391,143 @@ function _give_show_test_mode_notice_in_admin_bar( $wp_admin_bar ) {
...
 
 /**
+ * Add Link to Import page in from donation archive and donation single page
+ *
+ * @since 1.8.13
+ */
+function give_import_page_link_callback() {
+	?>
+    <a href="<?php echo esc_url( give_import_page_url() ); ?>"
+       class="page-import-action page-title-action"><?php esc_html_e( 'Import Donations', 'give' ); ?></a>
```
Comment (ravinderk) [inline includes/admin/admin-actions.php]: @raftaar1191 unnecessary `esc_html`. Remove it.

Comment (raftaar1191) [reply includes/admin/admin-actions.php]: @ravinderk 
I was reading this article that why I used it. Please have a look into it
https://vip.wordpress.com/documentation/validating-sanitizing-escaping/#always-escape-late

File: includes/process-donation.php [hunk 15]
```diff
@@ -815,8 +815,10 @@ function give_register_and_login_new_user( $user_data = array() ) {
 	 */
 	do_action( 'give_insert_user', $user_id, $user_data );
 
-	// Login new user.
-	give_log_user_in( $user_id, $user_data['user_login'], $user_data['user_pass'] );
+	if ( true === (bool) apply_filters( 'give_log_user_in_on_register', true ) ) {
```
Comment (ravinderk) [inline includes/process-donation.php]: @raftaar1191 Add documentation to this filter

File: includes/process-donation.php [hunk 16]
```diff
@@ -891,12 +893,12 @@ function give_get_donation_form_user( $valid_data = array() ) {
...
-	$user['address']['state']   = ! empty( $_POST['card_state'] ) ? sanitize_text_field( $_POST['card_state'] ) : false;
-	$user['address']['country'] = ! empty( $_POST['billing_country'] ) ? sanitize_text_field( $_POST['billing_country'] ) : false;
-	$user['address']['zip']     = ! empty( $_POST['card_zip'] ) ? sanitize_text_field( $_POST['card_zip'] ) : false;
+	$user['address']['line1']   = ! empty( $_POST['card_address'] ) ? give_clean( $_POST['card_address'] ) : false;
+	$user['address']['line2']   = ! empty( $_POST['card_address_2'] ) ? give_clean( $_POST['card_address_2'] ) : false;
+	$user['address']['city']    = ! empty( $_POST['card_city'] ) ? give_clean( $_POST['card_city'] ) : false;
+	$user['address']['state']   = ! empty( $_POST['card_state'] ) ? give_clean( $_POST['card_state'] ) : false;
+	$user['address']['zip']     = ! empty( $_POST['card_zip'] ) ? give_clean( $_POST['card_zip'] ) : false;
+	$user['address']['country'] = ! empty( $_POST['billing_country'] ) ? ( ( $country_code = array_search( give_clean( $_POST['card_zip'] ), give_get_country_list() ) ) ? $country_code : give_clean( $_POST['card_zip'] ) ) : false;
```
Comment (ravinderk) [inline includes/process-donation.php]: @raftaar1191 Can you explain this

File: includes/admin/tools/class-settings-import.php [hunk 17]
```diff
@@ -0,0 +1,681 @@
...
+							'page'      => 'give-payment-history',
+						);
+						$text      = __( 'View Donations', 'give' );
+					}
+					?>
+					<?php
+					if ( ! empty( $report['duplicate_donor'] ) ) {
+						?>
+                        <p>
+							<?php echo esc_html( wp_sprintf( '%s duplicate donors detected', $report['duplicate_donor'] ), 'give' ); ?>
```
Comment (ravinderk) [inline includes/admin/tools/class-settings-import.php]: @raftaar1191 Please check this implementation. Fix them all.
screen shot 2017-08-29 at 2 28 56 pm

File: includes/admin/tools/class-settings-import.php [hunk 18]
```diff
@@ -0,0 +1,681 @@
...
+			$total   = (int) $_GET['total'];
+			$total   = $total - 1;
+			$success = (bool) $_GET['success'];
+			?>
+            <tr valign="top" class="give-import-dropdown">
+                <th colspan="2">
+                    <h2>
+						<?php
+						if ( $success ) {
+							echo sprintf( __( 'Import complete! %s donations imported', 'give' ), '<strong>' . $total . '</strong>' );
```
Comment (ravinderk) [inline includes/admin/tools/class-settings-import.php]: @raftaar1191 Use double quotes.
```
'<strong>' . $total . '</strong>'  --> "<strong>{$total}</strong>"
```

File: includes/admin/tools/class-settings-import.php [hunk 19]
```diff
@@ -0,0 +1,681 @@
...
+
+					<?php
+					if ( ! empty( $report['duplicate_donation'] ) ) {
+						?>
+                        <p>
+							<?php echo esc_html( wp_sprintf( '%s duplicate donations detected', $report['duplicate_donation'] ), 'give' ); ?>
+                        </p>
+						<?php
+					}
+					?>
```
Comment (ravinderk) [inline includes/admin/tools/class-settings-import.php]: @raftaar1191 Collect import report in an array and then print them once

File: includes/admin/tools/class-settings-import.php [hunk 20]
```diff
@@ -0,0 +1,681 @@
...
+            <script type="text/javascript">
+                jQuery(document).ready(function ($) {
+                    give_on_donation_import_start();
+                });
+            </script>
+			<?php
+		}
+
+		/**
+		 * Will return true if importing can be started or not else false.
```
Comment (ravinderk) [inline includes/admin/tools/class-settings-import.php]: @raftaar1191 Please add missing `@since` tag atleast

File: includes/admin/tools/class-settings-import.php [hunk 21]
```diff
@@ -0,0 +1,681 @@
...
+
+		/**
+		 * Print the Dropdown option for CSV.
+		 */
+		static function render_dropdown() {
+			$csv = (int) $_GET['csv'];
+			?>
+            <tr valign="top" class="give-import-dropdown">
+                <th colspan="2">
+                    <h2 id="give-import-title"><?php esc_html_e( 'Map CSV fields to donations', 'give' ) ?></h2>
```
Comment (ravinderk) [inline includes/admin/tools/class-settings-import.php]: @raftaar1191 remove unnecessary `esc_html` from hard coded strings

File: includes/admin/tools/class-settings-import.php [hunk 22]
```diff
@@ -0,0 +1,681 @@
...
+					$file = new SplFileObject( $file_dir, 'r' );
+					$file->seek( PHP_INT_MAX );
+					$total = $file->key() + 1;
+				}
+			}
+
+			return $total;
+		}
+
+		static function get_importer( $file_id, $index = 0, $delimiter = ',' ) {
```
Comment (ravinderk) [inline includes/admin/tools/class-settings-import.php]: @raftaar1191 Please add documentation of function if newly created

File: includes/admin/tools/class-settings-import.php [hunk 23]
```diff
@@ -0,0 +1,681 @@
...
+		/**
+		 * Is used to show the process when user upload the donor form.
+		 *
+		 * @since 1.2
+		 */
+		static function progress() {
+			$step = Give_Settings_Import::get_step();
+			?>
+            <ol class="give-progress-steps">
+                <li class="<?php echo( 1 === $step ? 'active' : '' ); ?>">Upload CSV file</li>
```
Comment (ravinderk) [inline includes/admin/tools/class-settings-import.php]: @raftaar1191 Is there any reason that these strings are not internationalize

File: includes/admin/tools/class-settings-import.php [hunk 24]
```diff
@@ -0,0 +1,681 @@
...
+
+		/**
+		 * Will return the import step.
+		 *
+		 * @since 1.2
+		 *
+		 * @return int $step on which step doest the import is on.
+		 */
+		static function get_step() {
+			$step    = (int) ( isset( $_REQUEST['step'] ) ? give_clean( $_REQUEST['step'] ) : false );
```
Comment (ravinderk) [inline includes/admin/tools/class-settings-import.php]: @raftaar1191 if accept the only int then why we set the default value to `false`. This should be `0`, right?

File: includes/admin/tools/data/class-give-tools-delete-import-donors.php [hunk 25]
```diff
@@ -0,0 +1,442 @@
...
+			}
+		}
+
+		return $page_donation_id[ $page ];
+	}
+
+	/**
+	 * Given a key, get the information from the Database Directly
+	 *
+	 * @since  1.8.12
```
Comment (ravinderk) [inline includes/admin/tools/data/class-give-tools-delete-import-donors.php]: @raftaar1191 is this wrong `@since` tag, please update them if any

File: assets/js/admin/admin-scripts.js (hunk 13 above)
Comment (raftaar1191) [reply assets/js/admin/admin-scripts.js]: It is being used to show the uploader tab when **add media button** is click

File: includes/process-donation.php (hunk 15 above)
Comment (raftaar1191) [reply includes/process-donation.php]: Done

File: includes/process-donation.php (hunk 16 above)
Comment (raftaar1191) [reply includes/process-donation.php]: Done

File: includes/admin/tools/class-settings-import.php (hunk 17 above)
Comment (raftaar1191) [reply includes/admin/tools/class-settings-import.php]: Done

File: includes/admin/tools/class-settings-import.php (hunk 19 above)
Comment (raftaar1191) [reply includes/admin/tools/class-settings-import.php]: Done

File: includes/admin/tools/class-settings-import.php (hunk 20 above)
Comment (raftaar1191) [reply includes/admin/tools/class-settings-import.php]: Done

File: includes/admin/tools/class-settings-import.php (hunk 21 above)
Comment (raftaar1191) [reply includes/admin/tools/class-settings-import.php]: @ravinderk 

I was reading this article that why I used it. Please have a look into it
https://vip.wordpress.com/documentation/validating-sanitizing-escaping/#always-escape-late

File: includes/admin/tools/class-settings-import.php (hunk 22 above)
Comment (raftaar1191) [reply includes/admin/tools/class-settings-import.php]: Done

File: includes/admin/tools/class-settings-import.php (hunk 23 above)
Comment (raftaar1191) [reply includes/admin/tools/class-settings-import.php]: Done

File: includes/admin/tools/data/class-give-tools-delete-import-donors.php (hunk 25 above)
Comment (raftaar1191) [reply includes/admin/tools/data/class-give-tools-delete-import-donors.php]: Done

File: includes/admin/tools/class-settings-import.php (hunk 24 above)
Comment (raftaar1191) [reply includes/admin/tools/class-settings-import.php]: @ravinderk did not get you, Can you please explain it again

File: includes/admin/tools/class-settings-import.php (hunk 18 above)
Comment (raftaar1191) [reply includes/admin/tools/class-settings-import.php]: Done

File: includes/admin/tools/class-settings-import.php (hunk 24 above)
Comment (raftaar1191) [reply includes/admin/tools/class-settings-import.php]: Got it. Fixed!

**Comment by DevinWalker** [comment]
@raftaar1191 Please see my response:

> WP User Creation
> In the CSV there is an option to enter Donor ID and Donor Email id both. If donor id is not empty then it will search for that ID in the user list. If found then all the details will be fetched from that user id or if that id does not exist then it will search for the email if that email is already registered then all the details from that email will be fetch and if that email is not being registered then it will register a new user with that email followed by first name and last name.

There should be _two_ separate columns in the CSV. Donor ID corresponds to the ID of the _Give_ donor. User ID corresponds to the ID of the _WordPress_ user. If user ID is provided and no donor ID, the donor is created and attached to that WP user. If a Donor ID exists and not a WP User, then the donor is first searched for, and if found then the details are used for that existing donor. If the donor is not found then the donor is created. Like wise with WP user. 

It would be _really_ great if the importer could log and output specific exceptions like above. For instance 

_"The importer could not find a donor with ID #234. New donor # 1345 was created."_

We can add this in a future version. Let's first focus on making it work right.

> Repeated Imports and Duplicates
> Solved, Now where a new donation is imported it will first check if this donation already exists or not. It will first check with the same Form id, price, author, and date as well. If all the match then it assumes to be the same donation and if not then

Great!

> Final Import "Report"
> Working on it ( Done 29.08.17 )

Thanks.

> Test vs Live Donations
> Now there is a column in the CSV to mention that the donation is ( test or live ) donation. The value should in a boolean value. If in the column no found then there is an option on the first stepto choose the type of donation

Great

> New Donors Donations Count
> Working on it ( Done 29.08.17 )

👍🏻

> Remove the "Imported" badge and Add "Imported on ..." Note
> Done, Added in donation single page as well in donor notes section
>
> Empty Note Created on Import
> Solved
> 
> CSS Issue with "Import Donations" button
> Solved
> 
> Delete Imported Data Tool Not Deleting All Data
> Working on it ( Done 29.08.17 )
> 
> Typos to Fix
> Fixed

👍🏻

> Media Uploader Modal
> Working on it

@raftaar1191 what's the status on this?

> Related to your comment point number 11
> 11. Media Uploader Modal
> ->It will be better to used ID in place of URL

Ok I guess it's not a big deal.

File: includes/admin/admin-actions.php (hunk 14 above)
Comment (DevinWalker) [reply includes/admin/admin-actions.php]: This is OK @raftaar1191

**Comment by raftaar1191** [comment]
@DevinWalker 

Solved all the issues and also added a column for User ID as well

File: includes/admin/admin-actions.php (hunk 14 above)
Comment (ravinderk) [reply includes/admin/admin-actions.php]: @raftaar1191 In case string is hard coded, not user defined

File: includes/admin/admin-actions.php [hunk 26]
```diff
@@ -391,9 +391,143 @@ function _give_show_test_mode_notice_in_admin_bar( $wp_admin_bar ) {
...
+    </style>
+	<?php
+}
+
+
+add_action( 'give_payments_page_top', 'give_import_page_link_callback', 11 );
+add_action( 'give_view_order_details_before', 'give_import_page_link_callback', 11 );
+
+
+function give_donation_import_callback() {
```
Comment (ravinderk) [inline includes/admin/admin-actions.php]: @raftaar1191 Add function description

File: includes/admin/admin-actions.php [hunk 27]
```diff
@@ -391,9 +391,143 @@ function _give_show_test_mode_notice_in_admin_bar( $wp_admin_bar ) {
...
+	$import_setting = array();
+	$fields         = isset( $_POST['fields'] ) ? $_POST['fields'] : null;
+
+	parse_str( $fields );
```
Comment (ravinderk) [inline includes/admin/admin-actions.php]: @raftaar1191 use `give_clean` to sanitize values

File: includes/admin/admin-actions.php (hunk 26 above)
Comment (raftaar1191) [reply includes/admin/admin-actions.php]: Done

File: includes/admin/admin-actions.php (hunk 27 above)
Comment (raftaar1191) [reply includes/admin/admin-actions.php]: Done

File: includes/admin/admin-actions.php (hunk 14 above)
Comment (raftaar1191) [reply includes/admin/admin-actions.php]: Done

**Comment by DevinWalker** [comment]
@raftaar1191 we're almost there! Please see the following:

## WP Users are being created without roles


See these users, without roles, are being created when I run an import. They should have the new Give Donor and WP's subscriber roles.

## Add Option to Delete CSV after import

If you upload a CSV with sensitive information to `wp-content/uploads/` it can be publicly accessible. Please add an option to delete the CSV after import for better security.

**Comment by raftaar1191** [comment]
@DevinWalker  working on the above

**Comment by raftaar1191** [comment]
@DevinWalker  Solved and add new featured

File: includes/misc-functions.php [hunk 28]
```diff
@@ -1274,9 +1274,644 @@ function give_set_upgrade_complete( $upgrade_action = '' ) {
...
+		'page'      => 'give-tools',
+		'tab'       => 'import',
+	);
+	$import_query_arg  = wp_parse_args( $parameter, $defalut_query_arg );
+
+	return add_query_arg( $import_query_arg, admin_url( 'edit.php' ) );
+}
+
+
+function give_save_import_donation_to_db( $raw_key, $row_data, $main_key = array(), $import_setting = array() ) {
```
Comment (ravinderk) [inline includes/misc-functions.php]: @raftaar1191 Missing function documentation

File: includes/misc-functions.php [hunk 29]
```diff
@@ -1274,9 +1274,644 @@ function give_set_upgrade_complete( $upgrade_action = '' ) {
...
+		$args['post_author'] = (int) $payment_data['user_info']['id'];
+	}
+
+	return $args;
+}
+
+/**
+ * Check if Import donation is duplicate
+ *
+ * Ref: https://xwp.github.io/engineering-best-practices/php/
```
Comment (ravinderk) [inline includes/misc-functions.php]: @raftaar1191 Why did you add this

File: includes/misc-functions.php [hunk 30]
```diff
@@ -1363,3 +1995,227 @@ function give_array_insert_after( $key, array &$array, $new_key, $new_value ) {
...
+ * @since 1.8.13
+ *
+ * @param int $customer_id Customer ID.
+ * @param string $key Optional. The meta key to retrieve. By default, returns data for all keys.
+ * @param bool $single Whether to return a single value.
+ *
+ * @return mixed Will be an array if $single is false. Will be value of meta data field if $single
+ *  is true.
+ */
+function get_customer_meta( $customer_id, $key = '', $single = false ) {
```
Comment (ravinderk) [inline includes/misc-functions.php]: @raftaar1191 User donor instead of customer

File: includes/misc-functions.php (hunk 28 above)
Comment (raftaar1191) [reply includes/misc-functions.php]: Done

File: includes/misc-functions.php (hunk 29 above)
Comment (raftaar1191) [reply includes/misc-functions.php]: The argument that is being passed to WP query will result in better performance of MYSQL

So I thought of putting a referral link from which source did I get this will be useful

Comment (ravinderk) [reply includes/misc-functions.php]: Remove this one.

File: includes/misc-functions.php (hunk 30 above)
Comment (raftaar1191) [reply includes/misc-functions.php]: Done

File: includes/misc-functions.php (hunk 28 above)
Comment (raftaar1191) [reply includes/misc-functions.php]: Done

File: includes/misc-functions.php (hunk 29 above)
Comment (raftaar1191) [reply includes/misc-functions.php]: Done

**Comment by DevinWalker** [comment]
## Importer Review 

#### The "Delete CSV after import" should be deleted by default



#### Move "Do not import" to the first position of the mapping dropdown



#### Use `_n()` to output appropriate language for singular usage 



#### Change to a button using `.button` WP core class



#### Fix bug where imported for levels all have default radio checked



#### Imported custom donations don't display "Custom" and created forms have "Custom" disabled incorrect

When a donation doesn't match a level it's marked as "Custom". I noticed these donations missing "Custom" and the form has "Custom Amount" to disabled incorrectly.

**Comment by raftaar1191** [comment]
@DevinWalker 

Fixed all the error and suggestion that you have added above

**Comment by DevinWalker** [comment]
Great job on this @raftaar1191

**Comment by raftaar1191** [comment]
Thanks @DevinWalker
